src/v/storage/disk_log_impl.cc

```diff
@@ -557,46 +558,56 @@ model::term_id disk_log_impl::term() const {
     }
     return _segs.back()->offsets().term;
 }
+static offset_stats empty_offsets(model::offset start_offset) {
+    offset_stats ret;
+    ret.start_offset = start_offset;
+    if (ret.start_offset > model::offset(0)) {
+        ret.dirty_offset = ret.start_offset - model::offset(1);
+        ret.committed_offset = ret.dirty_offset;
+    }
+    return ret;
+}
 
 offset_stats disk_log_impl::offsets() const {
-    if (_segs.empty()) {
-        offset_stats ret;
-        ret.start_offset = _start_offset;
-        if (ret.start_offset > model::offset(0)) {
-            ret.dirty_offset = ret.start_offset - model::offset(1);
-            ret.committed_offset = ret.dirty_offset;
-        }
-        return ret;
-    }
-    // NOTE: we have to do this because ss::circular_buffer<> does not provide
-    // with reverse iterators, so we manually find the iterator
-    segment_set::type end;
-    segment_set::type term_start;
-    for (int i = (int)_segs.size() - 1; i >= 0; --i) {
-        auto& seg = _segs[i];
-        if (!seg->empty()) {
-            if (!end) {
-                end = seg;
-            }
-            // find term start offset
-            if (seg->offsets().term < end->offsets().term) {
-                break;
-            }
-            term_start = seg;
-        }
-    }
-    if (!end) {
-        offset_stats ret;
-        ret.start_offset = _start_offset;
-        if (ret.start_offset > model::offset(0)) {
-            ret.dirty_offset = ret.start_offset - model::offset(1);
-            ret.committed_offset = ret.dirty_offset;
-        }
-        return ret;
+    /**
+     * We have to iterate backward over segments set as we have to find last
+     * non empty segment, last segment with initialized committed offset and
+     * first offset of current term
+     */
+    const auto rend = std::make_reverse_iterator(_segs.cbegin());
+    const auto rbegin = std::find_if(
+      std::make_reverse_iterator(_segs.cend()),
+      rend,
+      [](const segment_set::type& seg) { return !seg->empty(); });
+
+    // If all segments are empty, return empty offsets.
+    if (rend == rbegin) {
+        return empty_offsets(_start_offset);
     }
+
+    auto it_last_committed = std::find_if(
+      rbegin, rend, [](const segment_set::type& seg) {
+          return seg->offsets().committed_offset >= model::offset(0);
+      });
+    // If nothing was committed, use the first segment
+    const auto& last_committed = it_last_committed != rend ? *it_last_committed
+                                                           : _segs.front();
+
+    // find last segment with term that is smaller than current one
```

Comment:
my reading of std::lower_bound is that identifies an element that is `>=` to a particular value, so the comment here is a little confusing.

Reply:
That comment is probably my fault.

The assumption is that term will be found. If we have segments with terms `{5,5,5,6,6,<empty>}`, the "last" (greatest term value) non-empty term is 6. So the search is for the first item that is not less than 6, or, the first item that is >= 6. That is index 3.

Reply:
cool. I guess we should fix the comment